server-common/src/main/java/org/apache/kafka/server/metrics/KafkaYammerMetrics.java

```diff
@@ -53,16 +72,21 @@ public static MetricsRegistry defaultRegistry() {
     }
 
     private final MetricsRegistry metricsRegistry = new MetricsRegistry();
-    private final FilteringJmxReporter jmxReporter = new FilteringJmxReporter(metricsRegistry,
-        metricName -> true);
+    private FilteringJmxReporter jmxReporter;
 
     private KafkaYammerMetrics() {
-        jmxReporter.start();
-        Exit.addShutdownHook("kafka-jmx-shutdown-hook", jmxReporter::shutdown);
     }
 
     @Override
+    @SuppressWarnings("deprecation")
     public void configure(Map<String, ?> configs) {
+        AbstractConfig config = new AbstractConfig(CONFIG_DEF, configs);
+        List<String> reporters = config.getList(CommonClientConfigs.METRIC_REPORTER_CLASSES_CONFIG);
+        if (config.getBoolean(CommonClientConfigs.AUTO_INCLUDE_JMX_REPORTER_CONFIG) || reporters.stream().anyMatch(r -> JmxReporter.class.getName().equals(r))) {
```

Comment:
Here's another possible call site for `shouldAddJmxReporter`.

Reply:
It's not trivial to consolidate all call sites as they are slightly different.

This instance and the one in `DynamicBrokerConfig` use `Map` objects for configs. In `DynamicBrokerConfig`, reporters are in the Scala `Buffer` instead of Java `List`. And finally depending on the file, the reporters are either class names or class instances.

Reply:
OK, well thanks for trying.